### ISSUE TITLE
open replication of the prior

### ISSUE THREAD

**Issue by rom1504** [lucidrains/DALLE2-pytorch#23]
hey,

thanks @lucidrains for building this awesome replication of the model, as usual!
dalle2 paper https://arxiv.org/pdf/2204.06125.pdf

with a few people from laion we're working on replication of the prior at scale
we're gathering notes in https://docs.google.com/document/d/1BKIQPzZS7pVL2JgL74W0dUIlUcfld8jptA6nIne_cNo/edit
and here
big plan:
* [x] use https://github.com/rom1504/embedding-reader to build a data loader
  * [x] use laion2B-en 2B image/text embeddings https://mystic.the-eye.eu/public/AI/cah/laion5b/embeddings/laion2B-en/img_emb/
* [x] plug that into the diffusion prior here: need to support precomputed embeddings
* [x] train at small scale
* [x] implement metrics (eg retrieval, classification, generation) on the output of the prior
* [x] train at large scale

Anybody interested in that project, feel free to discuss here or in #dalle2-prior at laion server (https://discord.gg/xBPBXfcFHd)

hp param from dalle2 paper
![image](https://user-images.githubusercontent.com/2346494/164995465-79f1e6e6-5aba-415f-918d-e349ac5119cd.png)

As a first step, we're trying to have an end to end version running at small scale. Once it works, we'll scale it up.

We intend to send PR in this repo for any improvement that seems worthwhile (eg support precomputed embeddings, distributed training, ...)

**Comment by lucidrains** [comment]
@rom1504 ah looks like a great plan :) so i think the PCA portion was only for the autoregressive prior method, and not the diffusion prior (they needed to quantize so they can use the straightforward cross entropy loss). but correct me if i'm wrong i'm happy to build that into the framework

**Comment by rom1504** [comment]
ohhh you're right, well that makes thing much easier
I'll remove that from the plan

**Comment by rom1504** [comment]
I'll keep this issue updated as we make progress

**Comment by lucidrains** [comment]
![6dp8bj](https://user-images.githubusercontent.com/108653/164996908-097ea317-4928-4bb8-a3a1-434a1eb46fb2.jpg)

**Comment by nousr** [comment]
I was able to over-fit a small subset of LAION-2B locally (using the new CLIP-less DiffusionPrior class). I'll be working to make that data loader next so we can do the first "real" training run. 

_referencing_:
>  - [ ] use https://github.com/rom1504/embedding-reader to build a data loader
>     - [ ] use laion2B-en 2B image/text embeddings https://mystic.the-eye.eu/public/AI/cah/laion5b/embeddings/laion2B-en/img_emb/

**Comment by rom1504** [comment]
Cool!
The corresponding text embeddings are now in https://mystic.the-eye.eu/public/AI/cah/laion5b/embeddings/laion2B-en/text_emb/

**Comment by rom1504** [comment]
You can use 2 instances of embedding reader and a zip to get both embeddings batch

**Comment by taki0112** [comment]
@rom1504 
The text embedding you mentioned is from which network? Is it a clip?

**Comment by rom1504** [comment]
Yes, image and text embeddings link above are ViT-L/14 clip.
Check the laion5B blogpost for details

**Comment by rom1504** [comment]
https://huggingface.co/rom1504/dalle2-diffusion-prior/resolve/main/1651432174.5708027_saved_model.pth here's a first checkpoint by @krish240574 
thanks to him for building the training code and running that first training!

it's time for the evaluation to start.

**Comment by krish240574** [comment]
> https://huggingface.co/rom1504/dalle2-diffusion-prior/resolve/main/1651432174.5708027_saved_model.pth here's a first checkpoint by @krish240574 thanks to him for building the training code and running that first training!
> 
> it's time for the evaluation to start.

Here is a run happening, 100 million data points - 
https://wandb.ai/laion/diffusion-prior/runs/3o0ic6ou?workspace=user-krish240574 

Hyperparameters - 
learning_rate=0.001,
max_grad_norm=0.5,
weight_decay=0.01
batch_size=10 ** 4
Train-val-test = 0.7 - 0.2 - 0.1 
Refer to https://github.com/lucidrains/DALLE2-pytorch/blob/main/train_diffusion_prior.py for training details.

**Comment by lucidrains** [comment]
@krish240574 so just a word of caution, that learning rate of `1e-3` is quite high for transformers. a more conservative value would be Karpathy's favorite, `3e-4` (however, I don't know how transformers behave within a DDPM framework, so I could be wrong)

**Comment by rom1504** [comment]
@krish240574 is running that new run with the newly added metrics https://wandb.ai/laion/diffusion-prior/runs/aul0rhv5?workspace=user-rom1504

seems cosine similarity is going up.

@lucidrains do you have any opinion on what would be the best way to know if the prior is doing its job? (except for plugging it into a generator and training of course)

**Comment by xiankgx** [comment]
Seems like the prior is doing a great job here. Please share the training details and checkpoint and dataset if possible. Thanks.

**Comment by rom1504** [comment]
Dataset is 
parser.add_argument("--image-embed-url", type=str, default="https://mystic.the-eye.eu/public/AI/cah/laion5b/embeddings/laion2B-en/img_emb/")
    parser.add_argument("--text-embed-url", type=str, default="https://mystic.the-eye.eu/public/AI/cah/laion5b/embeddings/laion2B-en/text_emb/")

IE laion2B-en clip ViT-l/14 text/image embeddings

**Comment by krish240574** [comment]
Here is a sample run, 600 million data points - https://wandb.ai/laion/diffusion-prior/runs/ar65uq6n?workspace=user-krish240574, hyperparameters as in the script(default values) train_diffision_prior.py

**Comment by nousr** [comment]
Just wanted to make a note that Katherine recommended we use EMA while training the prior as well @lucidrains

**Comment by lucidrains** [comment]
> Just wanted to make a note that Katherine recommended we use EMA while training the prior as well @lucidrains

ohh got it, i can take care of that tomorrow morning 👍

**Comment by lucidrains** [comment]
> Just wanted to make a note that Katherine recommended we use EMA while training the prior as well @lucidrains

https://github.com/lucidrains/DALLE2-pytorch/commit/740d644050687b1b09f2e22fce056b8ad3652bd5 ok all done! @crowsonkb thank you for the advice yet again :pray:

**Comment by lucidrains** [comment]
@nousr no problem! :)

yea we can aim for maybe the same size as GPT2 small? https://huggingface.co/docs/transformers/model_doc/gpt2

so it would translate to

```python
prior_network = DiffusionPriorNetwork(
    dim = 768,
    depth = 12,
    dim_head = 64,
    heads = 12
).cuda()
```

**Comment by lucidrains** [comment]
in light of https://github.com/lucidrains/DALLE2-pytorch/issues/71 should definitely be retrained at the latest version! i've also turned off classifier free guidance, since i'm uncertain if it works well with the predict x0 objective. final thought is that we should be training with the text encodings + corresponding text mask if possible. i think the paper did have this (although if Katherine was able to get good results without, let us aim for that first)

**Comment by crowsonkb** [comment]
Classifier-free guidance works fine with predicting x_0, I trained mine that way. :) As for training with the text encodings, I am feeding in the hidden states at the end of the frozen CLIP text encoder to my prior (along with the corresponding padding mask for attention) instead of trying to learn language from scratch. This works pretty well, a lot better than trying to feed in only the CLIP text embedding!

**Comment by crowsonkb** [comment]
Actually do you think I need learned queries if I am feeding in a sequence of text encoder hidden states? I don't have them and don't know how much they help. There are always at least two text encoder hidden states from the SOT and EOT tokens (for the null condition), more for actual prompts.

**Comment by lucidrains** [comment]
> Classifier-free guidance works fine with predicting x_0, I trained mine that way. :) As for training with the text encodings, I am feeding in the hidden states at the end of the frozen CLIP text encoder to my prior (along with the corresponding padding mask for attention) instead of trying to learn language from scratch. This works pretty well, a lot better than trying to feed in only the CLIP text embedding!

thanks for sharing! that's great to know! was wondering if this would work well given the l2norm constraint - i'll revert the commit next week so others can use it

and yes, i have things setup exactly like you did (the text encodings being the output of the final layer of the CLIP text transformer), so it should be ready to go, provided Laion can save and dataload the text encodings efficiently in addition to the text embeddings

**Comment by lucidrains** [comment]
> Actually do you think I need learned queries if I am feeding in a sequence of text encoder hidden states? I don't have them and don't know how much they help. There are always at least two text encoder hidden states from the SOT and EOT tokens (for the null condition), more for actual prompts.

so the literature is scant on this, but https://arxiv.org/abs/2006.11527 does suggest adding learned queries (up to 16) can be beneficial. i was also puzzled why they didn't take the output token of the image embedding, and one possibility is that they projected the "noised | predicted" image embedding into multiple tokens for attention (which i'll build into the repository as a setting eventually). the more realistic answer is no matter which token you choose, as long as the position information is there, and as long as the transformer is big enough, it won't matter that much :laughing: 

long answer short, it doesn't hurt to add a few memory (learned query) tokens. with a big enough transformer, probably matters little

**Comment by crowsonkb** [comment]
> i was also puzzled why they didn't take the output token of the image embedding

I tried this and it was worse/didn't learn as well. I suspect it might be easier to predict a clean x_0 without the residual bias from the noisy input (because it would have to generate exact anti-noise in the ffns and add it to that token's residual stream), whereas if you have a separate output token you just have to copy information into it via attention.

**Comment by lucidrains** [comment]
this is valuable information! thanks!

at the end of the day, I think the team still struggled with the generator not treating the text as a bag of words, and then bottleneck is the clip text encoder. I suspect the new Coca, which is an LM and Clip trained end to end, would help

**Comment by crowsonkb** [comment]
I think so too because it can't just throw away as much information about the contents of the individual tokens bc it needs to do the next token prediction task too.

**Comment by crowsonkb** [comment]
btw I think CoCa should be trained dropping out the image features so you can generate captions with superconditioning (it should work well there for the same reason it works well when generating images, the typical image/caption pair in most training sets doesn't match all that well).

**Comment by lucidrains** [comment]
> btw I think CoCa should be trained dropping out the image features so you can generate captions with superconditioning (it should work well there for the same reason it works well when generating images, the typical image/caption pair in most training sets doesn't match all that well).

yes! agreed on the superconditioning!

last thought i've had that is worth sharing is perhaps there can be one more level of indirection. If one were to train CLIP with a multi-stage efficient ViT, we can do `text -> text embed + text encoding -> image embed + image encoding (~49-64 tokens) -> image`. in the ideal world, we have a 540B parameter PaLM + CoCa (with FILIP fine interactions between the text tokens and image tokens) + cascading DDPM generator of about 3B parameters. One can dream 😆

**Comment by lucidrains** [comment]
@rom1504 update: Romain reported that a research group out there have replicated the prior using the code in this repository for their CLIP generations

**Comment by lucidrains** [comment]
@nousr hey yea i did, only partly. but the general scaffold is there that it shouldn't take too much code to convert it to be config-based (by looking at the decoder training script and `dalle2_pytorch/train_config.py`)

this week i'm back to running around meeting talking with people and being generally unproductive, so feel free to jump in with a PR!

**Comment by lucidrains** [comment]
@nousr hey yea i did, only partly. but the general scaffold is there that it shouldn't take too much code to convert it to be config-based (by looking at the decoder training script and dalle2_pytorch/train_config.py)

this week i'm back to running around doing meeting, generally unproductive state of affairs; so feel free to jump in with a PR!

**Comment by rom1504** [comment]
I believe we got everything done here.
One last thing that would be helpful in my opinion would be a prior.md with these information:
* What is the prior and what can it do
* How to prepare a dataset for it
* How to run a training
* how to use a pretrained model
* What are good metrics
* how to plug into the decoder

I think that would close this topic.
I believe most of the information is already present in readme.md but having it in its dedicated file would help as the readme is getting very large now.

**Comment by rom1504** [comment]
pretty much done now
